src/openloco/localisation/stringmgr.cpp

```diff
@@ -1,23 +1,337 @@
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x4959A1, regs);
+    }
+
+    // TODO: decltype(value)
+    static void format_int(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495E2A, regs);
+    }
+
+    // TODO: decltype(value)
+    static void formatNumeric_4(uint16_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x4963FC, regs);
+    }
+
+    // TODO: decltype(value)
+    static void format_comma2dp32(uint32_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x4962F1, regs);
+    }
+
+    static void format_comma(uint16_t value, char* buffer)
+    {
+        registers regs;
+        regs.eax = (uint32_t) value;
+        regs.edi = (uint32_t) buffer;
+
+        call(0x495F35, regs);
+    }
+
+    // 0x004958C6
     void format_string(char* buffer, string_id id, void* args)
     {
         registers regs;
         regs.eax = id;
         regs.edi = (uint32_t)buffer;
         regs.ecx = (uint32_t)args;
-        call(0x004958C6, regs);
+
+        if (id >= 0x8000)
+        {
+            if (id < 0x8800)
+            {
+                // sub     id, 8000h
+                id -= 0x8000;
+
+                // add     args, 2
+                args = (uint8_t*) args + 2;
+
+                // imul    id, 20h
+                // id *= USER_STRING_SIZE;
+
+                // add     id, offset _userStrings
+                char* sourceStr = _userStrings[id];
+
+                // loc_4958EF:
+                // mov     dl, [id]
+                // mov     [buffer], dl
+                // inc     id
+                // inc     buffer
+                // or      dl, dl
+                // jnz     short loc_4958EF
+
+                // !!! TODO: original code is prone to buffer overflow.
+                buffer = strncpy(buffer, sourceStr, USER_STRING_SIZE);
+                buffer[USER_STRING_SIZE - 1] = '\0';
+
+                // dec     buffer
+                // TODO: ???
+
+                return;
+            }
+            else if (id < 41024) // 0xA040
+            {
+                // add     id, 0FFFF6119h
+                id -= 40679;
+
+                // movzx   args, word ptr [args]
+                uint16_t town_id = *(uint16_t*) args;
+                args = (uint8_t*) args + 2;
+
+                // imul    args, 270h
+                auto town = townmgr::get(town_id);
+
+                // lea     args, towns[args]
+                void* town_name = (void*) &town->name;
+
+                // call    format_string
+                format_string(buffer, id, town_name);
+
+                return;
+            }
+            else if (id == 41024)
```

Comment:
I'm not sure where this bit of code has come from.

Reply:
It's a translation of what's happening with the jump over at `0x4958D8`.

Reply:
Oh yes my bad. Thats an odd bit of code.